torch/lib/THD/master_worker/common/CommandChannel.hpp

```diff
@@ -14,12 +18,20 @@ struct MasterCommandChannel {
 
   bool init();
 
-  std::unique_ptr<rpc::RPCMessage> recvMessage(int rank);
   void sendMessage(std::unique_ptr<rpc::RPCMessage> msg, int rank);
 
 private:
+  std::tuple<rank_type, std::string> recvError();
+  void errorHandler();
+
   rank_type _rank;
   std::vector<int> _sockets;
+  std::unique_ptr<struct pollfd[]> _poll_events;
+
+  bool _started; // indicates whether error handler thread has started
+  std::atomic_bool _exiting; // informs error handler thread that we are exiting
+  std::unique_ptr<std::string> _error;
```

Comment:
Actually how about this solution: Open an unnamed pipe and use it in a 0th descriptor of poll. When the channel is destructed write to it, and this will wake up the error thread. Once it sees that it got some data there, it will know that it's an exit message.